astropy/table/bst.py

```diff
@@ -0,0 +1,614 @@
+        '''
+        rows = self.data.set_default(key, [])
+        rows.insert(np.searchsorted(rows, val), val)
+
+    def find(self, key):
+        '''
+        Find rows corresponding to the given key.
+        '''
+        return self.data.get(key, [])
+
+    def remove(self, key, data=None):
+        '''
+        Remove data from the given key.
+        '''
+        node = self.data.get(key, None)
+        if node is None or len(node) == 0:
+            return False
+        if data is None:
+            self.data.pop(key)
+            return True
+        if data not in node:
+            if len(node) == 0:
+                return False
+            raise ValueError("Data does not belong to correct node")
+        node.remove(data)
+        return True
+
+
+    def shift_left(self, row):
+        '''
+        Decrement rows larger than the given row.
+        '''
+        for key, node in self.data.items():
+            self.data[key] = [x - 1 if x > row else x for x in node]
+
+    def shift_right(self, row):
+        '''
+        Increment rows greater than or equal to the given row.
+        '''
+        for key, node in self.data.items():
+            self.data[key] = [x + 1 if x >= row else x for x in node]
+
+    def traverse(self):
+        '''
+        Return all nodes in this BST.
+        '''
+        l = []
+        for key, data in self.data.items():
+            n = Node(key, key)
+            n.data = data
+            l.append(n)
+        return l
+
+    def items(self):
+        '''
+        Return a list of key, data tuples.
+        '''
+        return [x for x in self.data.items() if len(x[1]) > 0]
+
+    def sort(self):
+        '''
+        Return a list of rows in order sorted by key.
+        '''
+        return [x for node in self.data.values() for x in node]
+
+    def range(self, lower, upper, bounds=(True, True)):
+        '''
+        Return row values in the given range.
+        '''
+        # we need Epsilon since bintrees searches for
+        # lower <= key < upper, while we might want lower <= key <= upper
+        # or similar
+        if not bounds[0]: # lower < key
+            lower = Epsilon(lower)
+        if bounds[1]: # key <= upper
+            upper = Epsilon(upper)
+        l = [v for v in self.data.value_slice(lower, upper)]
+        return [x for sublist in l for x in sublist]
+
+    def replace_rows(self, row_map):
+        '''
+        Replace rows with the values in row_map.
+        '''
+        for data in self.data.values():
+            data[:] = [row_map[x] for x in data if x in row_map]
+
+    def __str__(self):
+        return str(self.data)
+
+    def __repr__(self):
+        return str(self)
+
+try:
+    # bintrees is an optional dependency
+    from bintrees import FastBinaryTree, FastRBTree
+
+    class FastBST(FastBase):
+        engine = FastBinaryTree
+
+    class FastRBT(FastBase):
+        engine = FastRBTree
+
+except ImportError:
+    FastBST = BST
+    FastRBT = BST
```

Comment:
Ah, perhaps all the more reason to put a generic binary tree module somewhere in Astropy, that we can both use (so that we can also take advantage of the optional bintrees dependency wherever applicable).